velox/exec/prefixsort/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,343 @@
+/// and memcpy operators, means not ValueSwappable.
+/// This is also the reason we can not use std::sort directly, and implement
+/// quickSort in PrefixSortContext.
+struct PrefixSortIterator {
+  PrefixSortIterator(char* prefix, const uint64_t entrySize)
+      : prefix_(prefix), entrySize_(entrySize) {}
+
+  PrefixSortIterator(const PrefixSortIterator& other)
+      : prefix_(other.prefix_), entrySize_(other.entrySize_) {}
+
+  // Inline functions should be place in include file to avoid link error.
+  inline char* operator*() const {
+    return prefix_;
+  }
+
+  inline PrefixSortIterator& operator++() {
+    prefix_ += entrySize_;
+    return *this;
+  }
+
+  inline PrefixSortIterator& operator--() {
+    prefix_ -= entrySize_;
+    return *this;
+  }
+
+  inline PrefixSortIterator operator++(int) {
+    auto tmp = *this;
+    prefix_ += entrySize_;
+    return tmp;
+  }
+
+  inline PrefixSortIterator operator--(int) {
+    auto tmp = *this;
+    prefix_ -= entrySize_;
+    return tmp;
+  }
+
+  inline PrefixSortIterator operator+(const uint64_t i) const {
+    auto result = *this;
+    result.prefix_ += i * entrySize_;
+    return result;
+  }
+
+  inline PrefixSortIterator operator-(const uint64_t i) const {
+    PrefixSortIterator result = *this;
+    result.prefix_ -= i * entrySize_;
+    return result;
+  }
+
+  inline PrefixSortIterator& operator=(const PrefixSortIterator& other) {
+    VELOX_CHECK_EQ(entrySize_, other.entrySize_);
+    prefix_ = other.prefix_;
+    return *this;
+  }
+
+  inline uint64_t operator-(const PrefixSortIterator& other) const {
+    VELOX_CHECK_EQ((prefix_ - other.prefix_) % entrySize_, 0);
+    VELOX_CHECK_GE(prefix_ - other.prefix_, 0);
+    return (prefix_ - other.prefix_) / other.entrySize_;
+  }
+
+  inline bool operator<(const PrefixSortIterator& other) const {
+    return prefix_ < other.prefix_;
+    ;
+  }
+
+  inline bool operator>(const PrefixSortIterator& other) const {
+    return prefix_ > other.prefix_;
+    ;
+  }
+
+  inline bool operator>=(const PrefixSortIterator& other) const {
+    return prefix_ >= other.prefix_;
+    ;
+  }
+
+  inline bool operator<=(const PrefixSortIterator& other) const {
+    return prefix_ <= other.prefix_;
+    ;
+  }
+
+  inline bool operator==(const PrefixSortIterator& other) const {
+    return prefix_ == other.prefix_;
+    ;
+  }
+
+  inline bool operator!=(const PrefixSortIterator& other) const {
+    return prefix_ != other.prefix_;
+  }
+
+ private:
+  char* prefix_;
+  const uint64_t entrySize_;
+};
+
+/// Provides a series of methods (mostly required by the sort
+/// algorithm) for PrefixSort. Maintains the buffer for swap operations and
+/// some necessary context information such as entrySize.
+class PrefixSortContext {
+ public:
+  PrefixSortContext(memory::MemoryPool* pool, uint64_t entrySize, char* end);
+
+  inline void swap(const PrefixSortIterator& lhs, const PrefixSortIterator& rhs)
+      const {
+    VELOX_CHECK_LT(*lhs, end_);
```

Comment:
These checks are expensive. We usually use debug-only checks in hot loops. VELOX_DCHECK_XXX